app/models/piece.rb

```diff
@@ -4,8 +4,82 @@ class Piece < ActiveRecord::Base
 
 	enum status: [:black, :white]
 
-  def is_obstructed(destination_x, destination_y)
+  def is_obstructed?(destination_x, destination_y)
+    x1, y1 = self.x, self.y
+    x2, y2 = destination_x, destination_y
 
+    if x2 == x1 #vertical line
+      x_distance = 0
+    else        #diagonal/horizontal line
+      if x1 > x2
+        x_distance = -((x2..x1).to_a.length - 1)
+      else
+        x_distance = ((x1..x2).to_a.length) - 1
+      end
+    end
+
+    slope = (y2 - y1).to_f / (x_distance).to_f
+
+    if slope == 1 || slope == -1
+      diagonal_obstruction?(x2, y2)
+    elsif slope == 0
+      horizontal_obstruction?(x2, y2)
+    elsif slope == -Float::INFINITY||
+          slope == Float::INFINITY
+      vertical_obstruction?(x2, y2)
+    else
+      invalid_move = "Invalid input. Not diagonal, horizontal, or vertical"
+      raise invalid_move
+    end
+  end
+
+  def set_x_coordinates(destination_x)
+    if destination_x > self.x
+      x_coordinates = (self.x...destination_x).to_a[1..-1]
+    else
+      x_coordinates = (destination_x...self.x).to_a.reverse[0..-2]
+    end
+  end
+
+  def set_y_coordinates(destination_y)
+    if destination_y > self.y
+      y_coordinates = (self.y...destination_y).to_a[1..-1]
+    else
+      y_coordinates = (destination_y...self.y).to_a.reverse[0..-2]
+    end
+  end
+
+  def check_coordinates(x_coordinates, y_coordinates)
```

Comment:
Yeah its checking both x and y separately because when I have a vertical obstruction, I don't need to check for both x and y, only for y, because x will be the same for all points(they lie on the same vertical plane). What do you think?